src/libraries/filcrypto/filproofs/algorithms.go

```diff
@@ -266,62 +268,59 @@ func (sdr *WinStackedDRG_I) Seal(sid sector.SectorID, data []byte, randomness se
 		CommCTreePath_:     commCTreePath,
 		CommQTreePath_:     commQTreePath,
 		CommRLastTreePath_: commRLastTreePath,
-		Seeds_:             sealSeeds,
+		Seed_:              sealSeed,
 		KeyLayers_:         windowKeyLayers,
 		QLayer_:            qLayer,
 		Replica_:           replica,
 	}
 	return &result
 }
 
-func (sdr *WinStackedDRG_I) _generateWindowKey(windowIndex int, sid sector.SectorID, commD sector.UnsealedSectorCID, nodes int, randomness sector.SealRandomness) ([][]byte, sector.SealSeed) {
-	sealSeed := computeSealSeed(sid, windowIndex, randomness, commD)
+func (sdr *WinStackedDRG_I) _generateWindowKey(sealSeed sector.SealSeed, windowIndex int, sid sector.SectorID, commD sector.UnsealedSectorCID, nodes int, randomness sector.SealRandomness) [][]byte {
 	nodeSize := int(sdr.NodeSize())
 	curveModulus := sdr.Curve().FieldModulus()
 	layers := int(sdr.Layers())
 
 	keyLayers := generateSDRKeyLayers(sdr.Drg(), sdr.Expander(), sealSeed, windowIndex, nodes, layers, nodeSize, curveModulus)
 
-	return keyLayers, sealSeed
+	return keyLayers
 }
 
 func (sdr *WinStackedDRG_I) GenerateCommitments(replica []byte, windowKeyLayers [][]byte, qLayer []byte) (commC PedersenHash, commQ PedersenHash, commRLast PedersenHash, commR PedersenHash, commCTreePath file.Path, commQTreePath file.Path, commRLastTreePath file.Path) {
 	commC, commCTreePath = computeCommC(windowKeyLayers, int(sdr.NodeSize()))
 	commQ, commQTreePath = computeCommQ(qLayer, int(sdr.NodeSize()))
 	commRLast, commRLastTreePath = BuildTree_PedersenHash(replica)
-
 	commR = TernaryHash_PedersenHash(commC, commQ, commRLast)
 
-	// FIXME: need to compute commQ.
 	return commC, commQ, commRLast, commR, commCTreePath, commQTreePath, commRLastTreePath
 }
 
-func computeSealSeed(sid sector.SectorID, windowIndex int, randomness sector.SealRandomness, commD sector.UnsealedSectorCID) sector.SealSeed {
-	var proverId []byte // TODO: Derive this from sid.MinerID()
+func getProverID(minerID sector.MinerID) []byte {
+	panic("TODO")
+}
+func computeSealSeed(sid sector.SectorID, randomness sector.SealRandomness, commD sector.UnsealedSectorCID) sector.SealSeed {
```

Comment:
We should remember to sync on this on spec completion and ensure all randomness generation is accounteed for together in the protocol (e.g. DSTs and the like). may be worth adding a TODO so we can find it later. Fair chance it doesn;t apply for randomness here but wanted to raise it in case.

Reply:
Randomness isn't being created here. Can you be more specific? Or even better, once we know exactly what we're trying to account for later, we can make a pass to identify relevant remaining points in here. The proofs code only ever consumes a few specific random values then uses them as our constructions dictate.

Reply:
I would argue the SealSeed is randomness, as in the whole preimage computation and hashing is creating the appropriate randomness value for you. @jzimmerman would push hard to use a CBOR-serialization of an object made up of the sector number window index, commD and proverID per #677 in trying to ensure we are consistent in generating randomness in a same way everywhere in the protocol. Is there a strong reason not to follow that pattern here? (I'm pretty on the fence about the net benefit of enforcing it throughout the protocol fwiw, but if we don't i'd want to weaken it elsewhere as well)...

Reply:
I am not sure I agree. I do not think we expect `SealSeed` to be random oracle output. As far as I know, it just needs to be unique. In the past, for example, we used pedersen hashing — which is not pseudorandom, just collision-resistant. This value is itself included in hashes to make their output unique. This operation, which we historically call 'KDF' does need to generate pseudorandom output, and the replicaID is effectively a nonce input.

I think it's important to distinguish between the 'random' values which are inputs to the proofs, and those which proofs per se generates as part of its internal workings. The former can be completely externalized, and the latter cannot be at all. If 'the protocol' (which I distinguish from proofs here, let's not quibble about whether that is meaningful) wants to take ownership of `SealSeed`, it would also need `CommD` for that purpose. I do *not* think `filproofs` should be calling into any general protocol code, nor should it have a dependency (conceptual or software-based on CBOR) — so this means that 'the protocol' would also need to take responsibility for calculating `CommD` — which in  turn will require it to handle preprocessing, etc. Given that `filproofs` depends on `CommD` having been constructed correctly, I think it should continue to own its construction. So this would require an extra library call such that 'whoever' is responsible for packing the sector would need to acquire `CommD`, then use it to generate `SealSeed` and pass it to proofs. This seems like a great deal of extra work, complexity, and places for things to go wrong — especially given that it's not necessary.

Moreover, if we're not prepared to accept that `filproofs` generates some kinds of 'randomness-like things' of its own accord, then we won't be able to perform the proofs at all. Once we accept that there is a qualitative distinction between the proofs inputs and values generated incidentally in the process of proving, then the problem (which I claim never existed) goes away. Much of proofs is necessarily a black box to the rest of the protocol.

Reply:
I think you're taking this way farther than I intended, and painting this as a very stark choice when it may be much simpler. 

All I'm saying is we should consider having proofs draw/generate randomness under the same structure that the reset of the Filecoin protocol does. Modulo how much work it is (it shouldn't be much imo save for the CBOR), it will help future implementers who may choose to architect the protocol differently. 

I don't think it's a big deal either way, but I hardly think it's tantamount to breaking the abstraction barriers chosen in order to share code modules or something. I'm just saying we could sample randomness the same way for consistency across all parts of FIL

Reply:
cc @jzimmerman . That said nbd if not someething proofs can do. It could well be that it does not rise to the level of urgency/necessity required for time to be spent on it under the current regimen. I just don't think the options are quite as stark as you present.

Reply:
As discussed offline, `SealSeed` (was `ReplicaID`) is not 'sampled randomness'. It is an identifier which need only be globally unique in combination with `SealCfg`.